### ISSUE TITLE
Questions about data parallel and memory usage

### ISSUE THREAD

**Issue by songcq** [microsoft/LightGBM#4619]
hi lightgbm team, thanks for your awesome work.

I have two questions about supporting bigger dataset. Please let me know if asking two questions in the same issue is not a good practice :)

### Summary
With data parallel, does lightgbm still merge global histogram from all local data? Or the histogram is generated using local data only?

### Motivation
If still using global data, then data parallel should produce as good models as non-parallel training.

If local data only, then the produced model will be less good, because the histogram doesn't have global information.

### Details
[the doc](https://lightgbm.readthedocs.io/en/latest/Features.html#data-parallel-in-lightgbm) mentioned that in traditional algorithm:

> - Partition data horizontally.
> - Workers use local data to construct local histograms.
> - Merge global histograms from all local histograms.

But in lightgbm

> Instead of “Merge global histograms from all local histograms”, LightGBM uses “Reduce Scatter” to merge histograms of different (non-overlapping) features for different workers. Then workers find the local best split on local merged histograms and sync up the global best split.

Is "Reduce Scatter" just a optimized way of merging global histograms? Thanks

### Summary
When max_bin is 255, how much memory does each feature consume per row?

### Details
In theory, when max_bin is 255, we could load 1 column of data, turn them into histogram, then remember of the range of each bin and drop the original data.

In that way each feature will consume much less memory. Does lightgbm do that?

Thanks

**Comment by jameslamb** [comment]
Thanks for using LightGBM! We can answer both questions here, but in the future please split multiple unrelated questions into multiple issues.

<hr>

> Is "Reduce Scatter" just a optimized way of merging global histograms

Maybe [this diagram in the NCCL docs](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/operations.html#reducescatter) will help you. It basically is like "after this operation completes, each worker holds a non-overlapping portion of the final result".

![image](https://user-images.githubusercontent.com/7608904/134041950-b40ba0d8-731e-44ef-a19d-ac2d14507797.png)

In LightGBM data-parallel training, every worker is responsible for evaluating splits on a subset of features in its local data.

Here's where LightGBM chooses the features relevant to each worker in `DataParallelTreeLearner::BeforeTrain()`:

https://github.com/microsoft/LightGBM/blob/0a847efe689239a3fea31b54b99159682e3ac989/src/treelearner/data_parallel_tree_learner.cpp#L60-L80

Then here's where, in `DataParallelTreeLearner::FindBestSplits()`, each worker evaluates all candidate splits for its subset of features, then performs an Allreduce sync up of those splits, so all workers can reach the same conclusion about the globally-best split.

https://github.com/microsoft/LightGBM/blob/0a847efe689239a3fea31b54b99159682e3ac989/src/treelearner/data_parallel_tree_learner.cpp#L174-L188

In case you're interested, you can find LightGBM's implementation of `ReduceScatter` and other such collective operations at https://github.com/microsoft/LightGBM/blob/d517ba12f2e7862ac533908304dddbd770655d2b/src/network/network.cpp

> If local data only, then the produced model will be less good, because the histogram doesn't have global information.

It's not exactly "local data only" with Data Parallel in LightGBM. Histogram boundaries for each feature are computed using all data across all workers, and those bins are are synced across the cluster to all workers.

https://github.com/microsoft/LightGBM/blob/0a847efe689239a3fea31b54b99159682e3ac989/src/treelearner/data_parallel_tree_learner.cpp#L130-L131

But yes, multi-worker data parallel training will produce different models than using equivalent parameters and data on a single machine. That's because the work of estimating the expect gain for different splits is parallelized across the cluster in a way that only one worker's data for a particular feature will be used to estimate the gain of splits for that feature.

@shiyu1994 @StrikerRUS please correct me if I've said anything incorrect above.

<hr>

> When max_bin is 255, how much memory does each feature consume per row?

I'm not sure about this question. Maybe other maintainers will be able to answer.

If you're comfortable reading C++ code, you can look at the interfaces for the `Dataset` (https://github.com/microsoft/LightGBM/blob/master/include/LightGBM/dataset.h) and `BinMapper` (https://github.com/microsoft/LightGBM/blob/d517ba12f2e7862ac533908304dddbd770655d2b/include/LightGBM/bin.h) classes.

> In theory, when max_bin is 255, we could load 1 column of data, turn them into histogram, then remember of the range of each bin and drop the original data. Does LightGBM do that?

Yes, that is how the LightGBM `Dataset` object works. If you're using wrapper packages like the R or Python package, you can pass parameter `free_raw_data = True` when creating a `Dataset` to achieve the behavior "clear the raw data from memory after constructing histograms".

https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.Dataset.html

**Comment by songcq** [comment]
Hi @jameslamb thanks for your reply :)

> Here's where LightGBM chooses the features relevant to each worker in `DataParallelTreeLearner::BeforeTrain()`:
> https://github.com/microsoft/LightGBM/blob/0a847efe689239a3fea31b54b99159682e3ac989/src/treelearner/data_parallel_tree_learner.cpp#L60-L80
> 
It seems there is no much randomness in this process? So a feature could be allocated to the same worker again and again? I might be missing something.

> But yes, multi-worker data parallel training will produce different models than using equivalent parameters and data on a single machine. That's because the work of estimating the expect gain for different splits is parallelized across the cluster in a way that only one worker's data for a particular feature will be used to estimate the gain of splits for that feature.

In theory it's possible to compute gain of splits of each feature using global data. LightGBM is not doing that to reduce communication cost?

With current design, we should be careful when distributing the data to each worker to make sure each subset is representative?

**Comment by shiyu1994** [comment]
For data parallel, the steps to find a best split threshold for a leaf is:
1. Each machine uses its local data to build local histograms.
2. Get the complete histograms for _all_ features on _all_ machines by `ReduceScatter` operation. The operation will sum up the local histograms from all machines to get the complete global histograms. Thus the global histograms are histograms summing from the whole dataset. Each machine will have a copy of these global histograms after `ReduceScatter`. Thus the comment 

>But yes, multi-worker data parallel training will produce different models than using equivalent parameters and data on a single machine. That's because the work of estimating the expect gain for different splits is parallelized across the cluster in a way that only one worker's data for a particular feature will be used to estimate the gain of splits for that feature.

is not accurate. Because the evaluation of splits only rely on the histograms. Since the histograms are built from the whole dataset, the evaluation of splits should be identical with training on a single machine.

3. Each machine is only responsible for finding the best split thresholds for a _subset_ of the features, from the _global_ histograms of these features.

To sum up, data parallel is data parallel only in the histogram construction stage. But after gathering all local histograms and getting the global histograms, the best split finding process is feature parallel, actually.

**Comment by songcq** [comment]
Hi @shiyu1994 thanks very much for your comments. 

> is not accurate. Because the evaluation of splits only rely on the histograms. Since the histograms are built from the whole dataset, the evaluation of splits should be identical with training on a single machine.
>  ...
> To sum up, data parallel is data parallel only in the histogram construction stage. But after gathering all local histograms and getting the global histograms, the best split finding process is feature parallel, actually.

So is it correct that the model produced by data parallel training should be as good as single machine training?

Thanks

**Comment by shiyu1994** [comment]
>So is it correct that the model produced by data parallel training should be as good as single machine training?

Yes. That's correct.

**Comment by jameslamb** [comment]
ah thanks so much for correcting me @shiyu1994 ! I learned a lot from your answers.

@songcq sorry for the incorrect information 😬

**Comment by songcq** [comment]
> @songcq sorry for the incorrect information
No worries, your links to the code are very helpful.
